### PR TITLE
Basic formatter

### PR DESCRIPTION
**Pull request by Kha**
- [x] Basic terms round-trip, more or less beautifully: https://github.com/leanprover/lean4/compare/master...Kha:formatter#diff-6173773b6ba5c12caa06815d79f81802R1-R35
- [x] Full modules round-trip, less beautifully: https://gist.github.com/Kha/91e55e8d75742a967d151c319e4205ea
- [ ] Add (soft and hard) newlines and indent at sensible places
- [ ] Optionally preserve existing formatting
- [ ] Use by default (given the import problem)

### CODE REVIEW

**Comment by Kha** [comment]
I've moved the commits not directly related to the formatter to the top of the commit list

File: src/Init/Util.lean [hunk 1]
```diff
@@ -13,6 +13,9 @@ universes u v
 def dbgTrace {α : Type u} (s : String) (f : Unit → α) : α :=
 f ()
 
+def dbgTraceVal {α : Type u} [HasToString α] (a : α) : α :=
```
Comment (Kha) [inline src/Init/Util.lean:16]: This is my new favorite function for debugging

File: src/Init/Data/String/Basic.lean [hunk 2]
```diff
@@ -490,6 +490,11 @@ if s.isNat then
 else
   none
 
+def beq (ss1 ss2 : Substring) : Bool :=
+ss1.toString == ss2.toString
```
Comment (Kha) [inline src/Init/Data/String/Basic.lean:494]: This could be sped up in a native implementation. Right now I only used it in tests, via `Syntax.structEq`.

**Comment by leodemoura** [comment]
merged

**Pull request closed without merging**